src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs

```diff
@@ -0,0 +1,176 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Net.Quic.Implementations.MsQuic.Internal;
+using System.Net.Security;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using static System.Net.Quic.Implementations.MsQuic.Internal.MsQuicNativeMethods;
+
+namespace System.Net.Quic.Implementations.MsQuic
+{
+    internal class MsQuicListener : QuicListenerProvider, IDisposable, IAsyncDisposable
+    {
+        // Functions to invoke in MsQuic
+        private MsQuicApi _api;
+
+        // Security configuration for MsQuic
+        private MsQuicSecurityConfig _secConfig;
+
+        // Pointer to the underlying listener
+        private IntPtr _nativeObjPtr;
+
+        // Handle to this object for native callbacks.
+        private GCHandle _handle;
+
+        // Delegate that wraps the static function that will be called when receiving an event.
+        private ListenerCallbackDelegate _listenerDelegate;
+
+        // Ssl listening options (ALPN, cert, etc)
+        private SslServerAuthenticationOptions _sslOptions;
+
+        private volatile bool _disposed;
+
+        private readonly Channel<MsQuicConnection> _acceptConnectionQueue = Channel.CreateBounded<MsQuicConnection>(new BoundedChannelOptions(512) // TODO make this configurable.
+        {
+            SingleReader = true,
+            SingleWriter = true
+        });
+
+        internal MsQuicListener(IPEndPoint listenEndPoint, SslServerAuthenticationOptions sslServerAuthenticationOptions, MsQuicApi api, IntPtr nativeObjPtr)
+        {
+            _api = api;
+            _sslOptions = sslServerAuthenticationOptions;
+            ListenEndPoint = listenEndPoint;
+            _nativeObjPtr = nativeObjPtr;
+
+            // TODO remove this.
+            StartAsync().GetAwaiter().GetResult();
```

Comment:
So a bit of discussion is required here, as this looks bad. 
@geoffkizer and I were discussing how to decide the states of the QuicListener here. Specifically, when creating the QuicListener, it would be nice if the listener has already started after newing it up. This effectively removes a normally required call to start the listener, as APIs like TcpListener pretty much always follow a pattern of calling ctor + Start() immediately after.

However, to accomplish this with MsQuic, before starting the listener, we really need to have a MsQuicSecurityConfig (or multiple). To create this security config in msquic, it is an asynchronous operation, therefore it can't really be called in the constructor without this sync over async.

I see a few options for approaches we can take to resolve this:
1. Add StartAsync like TcpListener which would resolve the security configurations. This would also match other apis.
2. Asking for msquic to provide a synchronous version of creating the security configuration, as I don't see a scenario where we call this API without an x509 certificate in hand.
3. Make the security config public, which needs to be resolved and passed into the QuicListener and QuicConnection.

Reply:
> Specifically, when creating the QuicListener, it would be nice if the listener has already started after newing it up.

I'm not sure I agree this is desirable. Generally I'd expect the constructor to set up the listener but not actually start listening until I call some API (`StartAsync`). That's how most of the networking APIs I've worked with behave.

Reply:
I agree that option 1 seems preferable.

Reply:
Also, if it's possible to stop and restart a listener, it will be necessary to have a `Start` API.

Reply:
I don't know how important making the listener restartable is. I also prefer option one as it matches other networking APIs as well.

Reply:
MsQuic does support restarting a listener. Also, it doesn't require a security config at start. It's explicitly design to allow for the app to dynamically go and load the sec config when the first request comes in, pend the NEW_CONNECTION event, and then set the sec config at a later time.

Reply:
> MsQuic does support restarting a listener.

Makes sense, the API surface made it seem like it was restartable. I just don't know 
> pend the NEW_CONNECTION event, and then set the sec config at a later time.

Oh interesting. this could be something worth exploring. This could enable the flow of calling start, and when we receive a NEW_CONNECTION, starting something async in the background to get the config. 

How do you got about starting the connection? I assume you set the security config on the connection via QuicSetParam?

Reply:
That's correct. There's a sec_config parameter you set on the connection.

Reply:
After setting the security_config, do you need to do anything to "unpend" the new connection, or does setting the security_config do that internally.

Reply:
Nope. That's it.

Reply:
I think there are a couple related issues here.

(1) Should the listener be restartable?
TcpListener supports this, but I really don't see how this is particularly useful. Why not just create a new listener? Maybe I'm missing something here.

(2) Should there be some sort of Start method (like e.g. TcpListener) or should creating the listener implicitly start it too?
Assuming the listener is not restartable, then we don't really *need* a Start method at all; instead, creating the listener can implicitly start this. I kind of like this because it's one less state for the listener to be in; instead of created/started/disposed it's just started/disposed. But I agree that this is different from the typical listener API, so perhaps we should add Start just for consistency.

(3) Should Start be async?
I don't think Start needs to be or should be async. I do understand that there's an implementation issue with msquic here. But I'd prefer to try to resolve that implementation issue first before making an API change. For now I think it's fine to do sync over async, and that's preferable to making an API change here that we're not sure we need.

Reply:
> (2) Should there be some sort of Start method (like e.g. TcpListener) or should creating the listener implicitly start it too?

Can start ever fail/throw? If so then it should not happen implicitly in the constructor.

Reply:
> Can start ever fail/throw? If so then it should not happen implicitly in the constructor.

It can, if we couldn't start the listener. However based on guidance I have heard, throwing in a constructor isn't the end of the world.

Reply:
> I don't think Start needs to be or should be async. I do understand that there's an implementation issue with msquic here. But I'd prefer to try to resolve that implementation issue first before making an API change. For now I think it's fine to do sync over async, and that's preferable to making an API change here that we're not sure we need.

I'm going to take a stab at resolving the security config later.

Reply:
However pending this investigation, a majority of people seem to favor having a StartAsync API as it matches other APIs still, so let's not rule out having StartAsync even if resolving the security config later may work.

Reply:
I had some trouble figuring out when to resolve the certificate for some methods Ex if OpenBidirectionalStream is called (which is sync), I would need to resolve the cert before creating the stream. Either way, I didn't want to do it in this PR as it will be tricky and I'd rather do it in follow up where it's clearer to see what is going on.

What do we need to do to make progress here? This seems to be only main blocking issue remaining as it affects public API.